### DISCUSSION TITLE
About DVC

### DISCUSSION THREAD

**Discussion by abaaba6666** [vincentjzy/OpenCorr#18] [Q&A] [answered]
First, I want to thank the developers for creating this awesome program!

I'm running into a problem with stacking 3D TIFF images. When I use the example TIFF in the Torus_tif folder, everything works perfectly. But when I try stacking all the TIFFs in the Torus folder the developers provide, the results are different and it looks like some data is missing.

Can someone help me understand how the stacking process works for these 3D TIFFs? Am I missing a step or setting something wrong? Also, after stacking, are the resulting TIFF files 64-bit or 32-bit?THANKS!
<img width="926" alt="image" src="https://github.com/user-attachments/assets/f65ded39-c470-4edf-88d7-cadc8d396877" />

**Comment by vincentjzy** [comment]
Please describe how you made your TIFF in details. It is hard to guess what happens to you according to the calculation results only. We provide two multi-layered TIFF of example Torus in opencorr.org. These are the volumetric images collected before and after deformation. There is no need to stack them again.

**Reply by abaaba6666** [reply]
Thanks to your reply dear developer!
In fact, I've used **Opencv** to stack multiple TIFF images from a folder and save them as an uncompressed multi-page TIFF file. （**My  code is pasted below**）;I've noticed that **_you've provided two multi-layered TIFF of example Torus in opencorr.org_**, and therefore I used the calculation results from the **_two example images on your website_** to **check** if the image stacking program I wrote is working correctly. But unfortunately, there seem to be something wrong with my stacking program. This is why I wanted to ask for your advice on how you stack the images.
Thanks a lot for your excellent development and for patiently answering my question!

（my code）
#include "opencv2/opencv.hpp"
#include <iostream>
#include <vector>
#include <string>
#include <windows.h> 

std::vector<std::string> get_tiff_files(const std::string& folder_path) {
    std::vector<std::string> tiff_files;

    // Define search path (including file type filter)
    std::string search_path = folder_path + "\\*.tif"; 
    WIN32_FIND_DATAA find_data; // Use A version
    HANDLE hFind = FindFirstFileA(search_path.c_str(), &find_data);

    if (hFind != INVALID_HANDLE_VALUE) {
        do {
            tiff_files.push_back(folder_path + "\\" + find_data.cFileName); 
        } while (FindNextFileA(hFind, &find_data) != 0); // Use A version
        FindClose(hFind);
    }

    return tiff_files;
}

int main() {
    // Input folder path
    std::string folder_path = "G:\\2025\\DVC explore\\Torus\\Torus\\ref";  // 
    std::string output_file = "G:\\2025\\DVC explore\\Torus\\Torus\\ref\\output_uncompressed.tif"; //

    // Get TIFF files in the folder
    std::vector<std::string> tiff_files = get_tiff_files(folder_path);
    if (tiff_files.empty()) {
        std::cerr << "No TIFF files found in folder: " << folder_path << std::endl;
        return -1;
    }

    // Load all images
    std::vector<cv::Mat> images;
    for (const auto& file : tiff_files) {
        cv::Mat img = cv::imread(file, cv::IMREAD_GRAYSCALE);  // Load image in grayscale mode
        if (img.empty()) {
            std::cerr << "Failed to load image: " << file << std::endl;
            return -1;
        }
        images.push_back(img);
    }

    
    std::vector<int> compression_params;
    compression_params.push_back(cv::IMWRITE_TIFF_COMPRESSION); 
    compression_params.push_back(1); // 1 means no compression

    // Save
    if (!cv::imwrite(output_file, images, compression_params)) {
        std::cerr << "Failed to save uncompressed multi-page TIFF file!" << std::endl;
        return -1;
    }

    std::cout << "Uncompressed multi-page TIFF file saved successfully: " << output_file << std::endl;

    return 0;
}

**Comment by vincentjzy** [comment, marked as answer]
You may find that the number of tiff files in folder Torus/def is less than that in folder Torus/ref, as the rubber ring is compressed. Thus, a few blank layers need to be inserted into the multi-layered tiff for def one. Otherwise the coordinate systems of the two volumetric images may not be properly aligned, as the origin of coordinate system is usually defined in the top layer.

**Reply by abaaba6666** [reply]
Wow, thank you so much! It's my first time doing the DVC calculation, and I didn’t realize how many things I missed. But your explanation just totally cleared things up for me! I'm so pumped and really appreciate your help!

**Comment by vincentjzy** [comment]
BTW, the check of dimension is implemented in OpenCorr GUI. If the dimensions of loaded images are different from the reference image, there will be a warning message.

**Reply by abaaba6666** [reply]
Thank you for the suggestion! I will further explore your program and learn from your approach.^_^
